### ISSUE TITLE
FxA Returning invalid verification code

### ISSUE THREAD

**Issue by jrbenny35** [mozilla/fxa#305]
This just started randomly on the [addons-server circleci](https://circleci.com/gh/mozilla/addons-server/tree/master).
Logs:
```
fxa_client = <fxa.core.Client object at 0x7fd87b317d50>
fxa_email = 'uitest@restmail.net'

@pytest.fixture
def fxa_account(fxa_client, fxa_email):
logger = logging.getLogger()
account = TestEmailAccount(email=fxa_email)
password = ''.join([random.choice(string.ascii_letters) for i in range(8)])
FxAccount = collections.namedtuple('FxAccount', 'email password')
fxa_account = FxAccount(email=account.email, password=password)
session = fxa_client.create_account(fxa_account.email,
fxa_account.password)
logger.info('Created: {}'.format(fxa_account))
account.fetch()
message = account.wait_for_email(lambda m: 'x-verify-code' in m['headers'])
> session.verify_email_code(message['headers']['x-verify-code'])

FxAccount = <class 'pytest_fxa.plugin.FxAccount'>
account = <fxa.tests.utils.TestEmailAccount object at 0x7fd87b442bd0>
fxa_account = FxAccount(email='uitest@restmail.net', password='hdIPrwgP')
fxa_client = <fxa.core.Client object at 0x7fd87b317d50>
fxa_email = 'uitest@restmail.net'
i = 7
logger = <logging.RootLogger object at 0x7fd88412ba90>
message = {'date': '2018-07-19T05:23:53.000Z', 'from': [{'address': 'verification@stable.dev.lcip.org', 'name': 'Firefox Account..."description" content="Verify your email to finish your Firefox Account registration"/>
</div>

</body>
</html>
', ...}
password = 'hdIPrwgP'
session = <fxa.core.Session object at 0x7fd87b5386d0>

.tox/ui-tests/local/lib/python2.7/site-packages/pytest_fxa/plugin.py:60:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <fxa.core.Session object at 0x7fd87b5386d0>
code = '8897ddd8454c260de07d60b7f85430db'

def verify_email_code(self, code):
> return self.client.verify_email_code(self.uid, code) # note: not authenticated

code = '8897ddd8454c260de07d60b7f85430db'
self = <fxa.core.Session object at 0x7fd87b5386d0>

.tox/ui-tests/local/lib/python2.7/site-packages/fxa/core.py:311:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <fxa.core.Client object at 0x7fd87b317d50>
uid = '2853d79b4ea3421c8fee2b837dd0b577'
code = '8897ddd8454c260de07d60b7f85430db'

def verify_email_code(self, uid, code):
body = {
"uid": uid,
"code": code,
}
url = "/recovery_email/verify_code"
> return self.apiclient.post(url, body)

body = {'code': '8897ddd8454c260de07d60b7f85430db', 'uid': '2853d79b4ea3421c8fee2b837dd0b577'}
code = '8897ddd8454c260de07d60b7f85430db'
self = <fxa.core.Client object at 0x7fd87b317d50>
uid = '2853d79b4ea3421c8fee2b837dd0b577'
url = '/recovery_email/verify_code'

.tox/ui-tests/local/lib/python2.7/site-packages/fxa/core.py:227:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <fxa._utils.APIClient object at 0x7fd87b317b90>
url = '/recovery_email/verify_code'
json = {'code': '8897ddd8454c260de07d60b7f85430db', 'uid': '2853d79b4ea3421c8fee2b837dd0b577'}
kwds = {}

def post(self, url, json=None, **kwds):
> return self.request("POST", url, json, **kwds)

json = {'code': '8897ddd8454c260de07d60b7f85430db', 'uid': '2853d79b4ea3421c8fee2b837dd0b577'}
kwds = {}
self = <fxa._utils.APIClient object at 0x7fd87b317b90>
url = '/recovery_email/verify_code'

.tox/ui-tests/local/lib/python2.7/site-packages/fxa/_utils.py:334:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <fxa._utils.APIClient object at 0x7fd87b317b90>, method = 'POST'
url = 'https://stable.dev.lcip.org/auth/v1/recovery_email/verify_code'
json = {'code': '8897ddd8454c260de07d60b7f85430db', 'uid': '2853d79b4ea3421c8fee2b837dd0b577'}
retry_auth_errors = True
kwds = {'headers': {'User-Agent': 'Mozilla/5.0 (Mobile; Firefox Accounts; rv:1.0) PyFxA/0.6.0 python-requests/2.19.1'}, 'timeout': 30}
resp = <Response [400]>
headers = {'User-Agent': 'Mozilla/5.0 (Mobile; Firefox Accounts; rv:1.0) PyFxA/0.6.0 python-requests/2.19.1'}
content_type = 'application/json; charset=utf-8'
body = {'code': 400, 'errno': 105, 'error': 'Bad Request', 'info': 'https://github.com/mozilla/fxa-auth-server/blob/master/docs/api.md#response-format', ...}

def request(self, method, url, json=None, retry_auth_errors=True, **kwds):
"""Make a request to the API and process the response.

This method implements the low-level details of interacting with an
FxA Web API, stripping away most of the details of HTTP. It will
return the parsed JSON of a successful responses, or raise an exception
for an error response. It's also responsible for backoff handling
and clock-skew tracking.
"""
# Don't make requests if we're in backoff.
# Instead just synthesize a backoff response.
if self._backoff_response is not None:
if self._backoff_until >= self.client_curtime():
resp = pickle.loads(self._backoff_response)
resp.request = None
resp.headers["Timestamp"] = str(int(self.server_curtime()))
return resp
else:
self._backoff_until = 0
self._backoff_response = None

# Apply defaults and perform the request.
while url.startswith("/"):
url = url[1:]
if not self.server_url.endswith("/"):
self.server_url = self.server_url + "/"
url = urljoin(self.server_url, url)
if self.timeout is not None:
kwds.setdefault("timeout", self.timeout)

# Configure the user agent
headers = kwds.get('headers', {})
headers.setdefault('User-Agent', USER_AGENT_HEADER)
kwds['headers'] = headers

resp = self._session.request(method, url, json=json, **kwds)

# Everything should return a valid JSON response. Even errors.
content_type = resp.headers.get("content-type", "")
if not content_type.startswith("application/json"):
msg = "API responded with non-json content-type: {0}"
raise fxa.errors.OutOfProtocolError(msg.format(content_type))
try:
body = resp.json()
except ValueError as e:
msg = "API responded with invalid json: {0}"
raise fxa.errors.OutOfProtocolError(msg.format(e))

# Check for backoff indicator from the server.
# If found, backoff up to the client-specified max time.
if resp.status_code in (429, 500, 503):
try:
retry_after = int(resp.headers["retry-after"])
except (KeyError, ValueError):
pass
else:
if self.max_retry_after is not None:
retry_after = max(retry_after, self.max_retry_after)
self._backoff_until = self.client_curtime() + retry_after
self._backoff_response = pickle.dumps(resp)

# If we get a 401 with "serverTime" field in the body, then we're
# probably out of sync with the server's clock. Check our skew,
# adjust if necessary and try again.
if retry_auth_errors:
if resp.status_code == 401 and "serverTime" in body:
try:
server_timestamp = int(body["serverTime"])
except ValueError:
msg = "API responded with non-integer serverTime: {0}"
msg = msg.format(body["serverTime"])
raise fxa.errors.OutOfProtocolError(msg)
# If our guestimate is more than 30 seconds out, try again.
# This assumes the auth hook will use the updated clockskew.
if abs(server_timestamp - self.server_curtime()) > 30:
self._clockskew = server_timestamp - self.client_curtime()
return self.request(method, url, json, False, **kwds)

# See if we need to adjust for clock skew between client and server.
# We do this automatically once per session in the hopes of avoiding
# having to retry subsequent auth failures. We do it *after* the retry
# checking above, because it wrecks the "were we out of sync?" check.
if self._clockskew is None and "timestamp" in resp.headers:
try:
server_timestamp = int(resp.headers["timestamp"])
except ValueError:
msg = "API responded with non-integer timestamp: {0}"
msg = msg.format(resp.headers["timestamp"])
raise fxa.errors.OutOfProtocolError(msg)
else:
self._clockskew = server_timestamp - self.client_curtime()

# Raise exceptions for any error responses.
# XXX TODO: hooks for raising error subclass based on errno.
if 400 <= resp.status_code < 500:
> raise fxa.errors.ClientError(body)
E           ClientError: Invalid verification code

body = {'code': 400, 'errno': 105, 'error': 'Bad Request', 'info': 'https://github.com/mozilla/fxa-auth-server/blob/master/docs/api.md#response-format', ...}
content_type = 'application/json; charset=utf-8'
headers = {'User-Agent': 'Mozilla/5.0 (Mobile; Firefox Accounts; rv:1.0) PyFxA/0.6.0 python-requests/2.19.1'}
json = {'code': '8897ddd8454c260de07d60b7f85430db', 'uid': '2853d79b4ea3421c8fee2b837dd0b577'}
kwds = {'headers': {'User-Agent': 'Mozilla/5.0 (Mobile; Firefox Accounts; rv:1.0) PyFxA/0.6.0 python-requests/2.19.1'}, 'timeout': 30}
method = 'POST'
resp = <Response [400]>
retry_auth_errors = True
self = <fxa._utils.APIClient object at 0x7fd87b317b90>
url = 'https://stable.dev.lcip.org/auth/v1/recovery_email/verify_code'

.tox/ui-tests/local/lib/python2.7/site-packages/fxa/_utils.py:320: ClientError
------------------------------ Captured log setup ------------------------------
plugin.py 57 INFO Created: FxAccount(email='uitest@restmail.net', password='hdIPrwgP')------------------------------- pytest-selenium --------------------------------
Driver log: /tmp/pytest-of-circleci/pytest-0/test_devhub_addon_upload_Resol0/driver.log URL: about:blank WARNING: Failed to gather log types: Message: GET /session/ec215684-0497-45a1-b137-34eb12f0ea41/log/types did not match a known command
```

**Comment by karlht** [comment]
I don't know the code well wnough to know if this is restmail.net returning an invalid code or the FxA verification failing to recognize a valid code.

Any ideas, FxA folks?

**Comment by vladikoff** [comment]
No idea at this time, when I saw this issue I tested prod right away and it was fine.

What env are you using for this ? @jrbenny35 @karlht ?

**Comment by shane-tomlinson** [comment]
I'm trying to reach out in the AMO channel to @jrbenny35 to see if this is still a problem.

**Comment by shane-tomlinson** [comment]
@jrbenny35 - from [this circle log](https://circleci.com/gh/mozilla/addons-server/21095), it appears that the same email address, uitest@restmail.net, is being used for every test. Right now there are 9 verification messages for that account, each with a distinct verification code: http://restmail.net/mail/uitest. If you are fetching the first verification email rather than the last, an incorrect verification code will be used and the 400 w/ errno 105 will be returned.

We typically use a distinct email address for every test we run to minimize email index issues like this. I'm not sure if this is the cause of your problem, but it stands out. I'm not sure [if the change from a mozilla.com to restmail.net address](I noticed that from https://github.com/mozilla/addons-server/pull/8559/files) could have anything to do with this.

**Comment by jrbenny35** [comment]
I use the pytest-fxa plugin. As seen [here](https://github.com/mozilla/pytest-fxa/blob/master/pytest_fxa/plugin.py#L71) it creates and destroys the email address. I don't think using the same address will be a problem unless someway an email didn't get destroyed correctly. Is there a way to check that?

**Comment by shane-tomlinson** [comment]
We have resolved this issue. These the things we noticed:

1. The test runner always uses the same email address, uitest@restmail.net
2. If the test runner fails creating & verifying an account, the [emails for that address are not deleted](https://github.com/mozilla/pytest-fxa/blob/master/pytest_fxa/plugin.py#L86).
3. PyFxA always [returns the first email that matches the predicate](https://github.com/mozilla/PyFxA/blob/97bfe7ad53650393e1b2dffa6786e0cc2072b090/fxa/tests/utils.py#L95), which in this case the predicate is whether an email has x-verify-code header.

Since the verify-code from the first email was always being returned, and then the auth-server was saying "invalid code", the emails for uitest were never deleted. A quick fix was to manually delete all of the emails for uitest.
